File: ETour-DotNet/Models/Booking.cs
namespace ETour.Api.Models;

public class Booking
{
    public int BookingId { get; set; }

    public int? UserId { get; set; }
    public User User { get; set; }

    public int? TourId { get; set; }
    public Tour Tour { get; set; }

    public int? ScheduleId { get; set; }
    public Schedule Schedule { get; set; }

    public DateOnly? BookingDate { get; set; }
    public decimal? TotalAmount { get; set; }
    public string BookingStatus { get; set; }

    public string ContactName { get; set; }
    public string ContactEmail { get; set; }
    public string ContactPhone { get; set; }

    public ICollection<PassengerDetails> Passengers { get; set; } = new List<PassengerDetails>();
    public ICollection<Payment> Payments { get; set; } = new List<Payment>();
    public Cancellation Cancellation { get; set; }
    public ICollection<Review> Reviews { get; set; } = new List<Review>();

    public string ResolveNotificationEmail()
    {
        if (!string.IsNullOrWhiteSpace(ContactEmail))
        {
            return ContactEmail.Trim();
        }
        return User?.Email;
    }
}

File: E-Tour/src/main/java/com/etour/entities/Booking.java
package com.etour.entities;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "booking")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Booking {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "booking_id")
    @EqualsAndHashCode.Include
    private Integer bookingId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    @ToString.Exclude
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "tour_id")
    @ToString.Exclude
    private Tour tour;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "schedule_id")
    @ToString.Exclude
    private Schedule schedule;

    @Column(name = "booking_date")
    private LocalDate bookingDate;

    @Column(name = "total_amount", precision = 12, scale = 2)
    private BigDecimal totalAmount;

    @Column(name = "booking_status", length = 30)
    private String bookingStatus;

    @Column(name = "contact_name", length = 120)
    private String contactName;

    @Column(name = "contact_email", length = 150)
    private String contactEmail;

    @Column(name = "contact_phone", length = 15)
    private String contactPhone;

    public String resolveNotificationEmail() {
        if (contactEmail != null && !contactEmail.isBlank()) {
            return contactEmail.trim();
        }
        return user != null ? user.getEmail() : null;
    }

    @OneToMany(mappedBy = "booking", fetch = FetchType.LAZY,
               cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    @ToString.Exclude
    private List<PassengerDetails> passengers = new ArrayList<>();

    @OneToMany(mappedBy = "booking", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @Builder.Default
    @ToString.Exclude
    private List<Payment> payments = new ArrayList<>();

    @OneToOne(mappedBy = "booking", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @ToString.Exclude
    private Cancellation cancellation;

    @OneToMany(mappedBy = "booking", fetch = FetchType.LAZY)
    @Builder.Default
    @ToString.Exclude
    private List<Review> reviews = new ArrayList<>();
}
